Tests/test_file_apng.py

```diff
@@ -663,6 +665,36 @@ def test_apng_save_blend(tmp_path):
         assert im.getpixel((0, 0)) == (0, 255, 0, 255)
 
 
+def test_save_all_progress():
+    out = BytesIO()
+    progress = []
+
+    def callback(filename, frame_number, n_frames):
+        progress.append((filename, frame_number, n_frames))
+
+    Image.new("RGB", (1, 1)).save(out, "PNG", save_all=True, progress=callback)
+    assert progress == [(None, 1, 1)]
```

Comment:
Maybe indicate index here (instead of `None`)? `1` for the 1st input image?

Reply:
Would you be open to the idea of having an index all of the time? Or, we could pass back the image instance, and let the user figure out the index or filename as they see fit.

Reply:
Having index always is probably simpler. As for passing the image, it wouldn't be much help with identifying the index if an image is reused in the input (like in that one test… come to think of it, having explicit index there would make adding a comment somewhat redundant).

Reply:
I've pushed a commit to switch to an index.

Reply:
...Why did you remove the filename though? I thought you were talking about _adding_ an index, not _removing_ the filename (as my own suggestion was to provide index _when the filename isn't available_)

Reply:
To clarify: can you add the filename back to the output? Indices are useful for ordering, but using actual filenames when available makes for much more user-friendly output (and it can be rather inconvenient to be forced to resolve filenames from input indices externally depending on usage scenario, while supplying appropriate filename along with the index internally would be trivial).

Reply:
I'm concerned that `(index, filename, frame_number, frame_total)` crosses the line from 'neat and tidy' to 'passing back everything to the user just in case something is useful'.

> it can be rather inconvenient to be forced to resolve filenames from input indices externally depending on usage scenario

I would have thought it was simple, given that the user just provided these images as the base image and `append_images`?

Reply:
> To clarify: can you add the filename back to the output?

Only `PIL.ImageFile.ImageFile` have filenames, while `append_images` are arbitrary `PIL.Image.Image`.

@radarhere why not pass just saved im object itself?

Reply:
I [suggested that](https://github.com/python-pillow/Pillow/pull/7435#discussion_r1342329935), but [it was pointed out that the image might be provided as an input more than once.](https://github.com/python-pillow/Pillow/pull/7435#discussion_r1342335519)

Reply:
…which is not the case with filenames, I suppose :-)

Reply:
I don't like passing indices since it's hard to figure out the difference between `index` and `frame_number` and it's very simple to write the code which uses `index` incorrectly.

Reply:
I'm not opposed to dropping `index` altogether - I think the key benefit of this feature is being able to track how many frames have been completed out of the total number.

Reply:
I'm still not convinced that a filename or image index is necessary, but I've pushed a commit to restore the filename.

```pycon
>>> from collections import namedtuple
>>> State = namedtuple("State", "a b")
>>> State(2, 4)[0]
2
```
Because namedtuples still have order, and I'm concerned that there could be different views about what the best order is, or that we might want to add more items in the future, I've used a dictionary instead.

Reply:
Either way, it would be a good idea to have this `getattr(im, "filename", None)` thing in one place in the codebase (and if you keep the `dict`, it would still be a good idea to have _one_ piece of code defining its creation, at least for sake of ensuring there won't be any discrepancies in key names).

Reply:
I've pushed a commit to move the creation of the dictionary into a common method.